kani-compiler/src/kani_middle/intrinsics.rs

```diff
@@ -0,0 +1,103 @@
+            if let TerminatorKind::Call { func, args, .. } = &mut terminator.kind {
+                let func_ty = func.ty(&self.local_decls, self.tcx);
+                if let Some((intrinsic_name, _args)) = resolve_rust_intrinsic(self.tcx, func_ty) {
+                    trace!(?func, ?intrinsic_name, "run_pass");
+                    if intrinsic_name == sym::simd_bitmask {
+                        self.replace_simd_bitmask(func, args)
+                    }
+                }
+            }
+        }
+    }
+
+    /// Change the function call to use the stubbed version.
+    /// We only replace calls if we can ensure the input has simd representation.
+    fn replace_simd_bitmask(&self, func: &mut Operand<'tcx>, args: &[Operand<'tcx>]) {
+        assert_eq!(args.len(), 1);
+        let tcx = self.tcx;
+        let arg_ty = args[0].ty(&self.local_decls, tcx);
+        if arg_ty.is_simd() {
+            // Get the stub definition.
+            let stub_id = tcx.get_diagnostic_item(Symbol::intern("KaniModelSimdBitmask")).unwrap();
+            debug!(?func, ?stub_id, "replace_simd_bitmask");
+
+            // Get SIMD information from the type.
+            let Operand::Constant(fn_def) = func else { unreachable!() };
+            let ty::FnDef(_, gen_args) = fn_def.literal.ty().kind() else { unreachable!() };
+            let (len, elem_ty) = simd_len_and_type(tcx, arg_ty);
+            debug!(?len, ?elem_ty, "replace_simd_bitmask Ok");
+
+            // Increment the list of generic arguments since our stub also takes element type and len.
+            let mut new_gen_args = Vec::from_iter(gen_args.iter());
+            new_gen_args.push(elem_ty.into());
+            new_gen_args.push(len.into());
+
+            fn_def.literal = ConstantKind::from_value(
+                ConstValue::ZeroSized,
+                tcx.type_of(stub_id).instantiate(tcx, &new_gen_args),
+            );
+        } else {
+            debug!(?arg_ty, "replace_simd_bitmask failed");
+        }
+    }
+}
+
+fn simd_len_and_type<'tcx>(tcx: TyCtxt<'tcx>, simd_ty: Ty<'tcx>) -> (Const<'tcx>, Ty<'tcx>) {
+    match simd_ty.kind() {
+        ty::Adt(def, args) => {
+            assert!(def.repr().simd(), "`simd_size_and_type` called on non-SIMD type");
+            let variant = def.non_enum_variant();
+            let f0_ty = variant.fields[0u32.into()].ty(tcx, args);
+
+            match f0_ty.kind() {
+                ty::Array(elem_ty, len) => (*len, *elem_ty),
+                _ => (Const::from_target_usize(tcx, variant.fields.len() as u64), f0_ty),
+            }
+        }
+        _ => unreachable!("unexpected layout for simd type {simd_ty}"),
+    }
+}
+
+fn resolve_rust_intrinsic<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    func_ty: Ty<'tcx>,
+) -> Option<(Symbol, GenericArgsRef<'tcx>)> {
+    if let ty::FnDef(def_id, args) = *func_ty.kind() {
+        if tcx.is_intrinsic(def_id) {
+            return Some((tcx.item_name(def_id), args));
```

Comment:
Why does this return `args`, if we are just ignoring it later and doing the same `FnDef` pattern match again manually in `replace_simd_bitmask`?

Reply:
I totally forgot that we had that already. Let me change that.